### DISCUSSION TITLE
Testing Approach

### DISCUSSION THREAD

**Discussion by Jaardy** [swe-delivery/swe-tube#16] [General]
Starting a new thread, as this is likely to generate a lot of discussion!

**Comment by Jaardy** [comment]
Matthew Gibbons
I've put unit tests close to the source code under test. This is quickly becoming good practice.

Matthew Gibbons
I've put spec tests (end-to-end tests) outside of the source tree, since they are testing the external interface.


Matthew Gibbons

https://kentcdodds.com/blog/colocation#tests

**Comment by accuser** [comment]
This is the approach that I tend gravitate towards: spec tests are for the product, unit tests are for me.

I've been using [Vitest](https://vitest.org) for testing more and more recently, particularly because it has support for TypeScript. It pretty much tracks Jest feature-wise, but with some additional some affordances. **I am not suggesting that we use Vitest** - it just informs some of my thinking.

The most interesting features it adds is _inline_ testing, or simply being able to co-locate a unit test _inside_ the module of interest. I find this interesting because there are some modern languages ([Rust](https://rustlang.org), for example) that have this sort of approach baked in. Personally, I don't think I could find a use for it with modules, because it breaks some of the principles of unit testing, although it might be useful with classes (but, I don't use classes).

One of the fundamentals of unit testing is that (you may have to pretend) you don't know how the code is implemented, you only know 'messages in, messages out'. In JavaScript, 'messages in' is described by the public interface of a class or the exports from a module, and 'messages out' can be thought of the same for the dependencies the class or module requires.

Placing the unit test for a class or module in an adjacent module (i.e., `module-name.test.js`) co-located in the same folder is A Good Thing&trade;: not only are the tests close to the subject of the tests, being external they need to access the module in the way it is intended to be used, i.e., `require` / `import` (I accept that there may be some barrelling going on elsewhere). We can then focus on unit tests that address messages in. Mocking is also reasonably straightforward, since we can mock the dependent modules in the unit tests (other options are available). The benefits of this approach over locating tests in a common `tests` folder are threefold: we don't have to worry about organising unit test modules, we don't have to maintain that organisation as code is refactored, and most importantly, unit tests are very easy to locate.

Integration tests (not full end-to-end) can be placed strategically. An endpoint handler might have an integration test that goes all the way to a mocked persistence layer, for example. This is testing the assemblage of parts; it makes sense to me that this is kept close to the endpoint handler module, as with the unit tests. I approach integration tests as unit testing bigger units.

Spec/behaviour tests are the full end-to-end tests. I think these should be kept separate. They require a lot more setup (the service / app under test could be running in production-like state, for example) and generally require external testing tools (e.g., [Playwright](https://playwright.dev)). I gravitate towards _spec_ tests because they are based on the specification agreed with the client; treating them separately is also A Good Thing&trade;.
